substrate/network/src/consensus.rs

```diff
@@ -137,11 +135,24 @@ impl Consensus {
 	}
 
 	pub fn on_bft_message(&mut self, io: &mut SyncIo, protocol: &Protocol, peer_id: PeerId, message: message::LocalizedBftMessage, hash: Hash) {
-		if self.messages.contains_key(&hash) {
+		if self.message_hashes.contains(&hash) {
 			trace!(target:"sync", "Ignored already known BFT message from {}", peer_id);
 			return;
 		}
 
+		match protocol.chain().block_status(&BlockId::Hash(message.parent_hash)) {
+			Err(e) => {
+				debug!(target:"sync", "Error reading blockchain: {:?}", e);
+				return;
+			},
+			Ok(status) => {
+				if status != BlockStatus::InChain {
+					trace!(target:"sync", "Ignored unknown parent BFT message from {}, hash={}", peer_id, message.parent_hash);
```

Comment:
this doesn't seem right.
  - we are building on parent `A`
  - our peers produce `B` and begin building on `B`. Due to latency we still think the chain head is A.
  - we shouldn't ignore messages about that consensus just because there was latency in delivering `B` to us.

Reply:
what we want to keep:
  - unrecognized `parent_hash` messages (for some time)
  - `parent_hash == chain head` messages

what we don't want to keep
  - `parent_hash` in chain but has children already